### PR TITLE
Fix sql failed when using replace function

### PR DESCRIPTION
**Pull request by EricZequan**
### What problem does this PR solve?

Issue Number: ref #9522 

Problem Summary:

### What is changed and how it works?

Tiflash only supports the replace function with the first parameter being of ColumnString type. 

So I added type judgment to `col_src` in the replace execution function and converted col_src into the `ColumnString` type that can be parsed correctly.

At the same time, I added the following functions to handle the corresponding Const examples:

- `vectorConstSrcAndReplace`- handle `replace(Const, Column, Const)`
- `vectorConstSrcAndNeedle`- handle `replace(Const, Const, Column)`
- `vectorConstSrc`- handle `replace(Const, Column, Column)`

The following are the test results of the relevant sql：
```sql
mysql> CREATE TABLE `tl7154cbef` (
    ->   `col_11` vector(1) NOT NULL,
    ->   `col_12` int(10) unsigned NOT NULL,
    ->   `col_13` vector NOT NULL,
    ->   `col_14` vector(3) NOT NULL,
    ->   VECTOR INDEX `idx_38_6` USING HNSW ((VEC_L2_DISTANCE(`col_11`)))
    -> ) ENGINE=InnoDB DEFAULT CHARSET=gbk COLLATE=gbk_chinese_ci;
Query OK, 0 rows affected (0.04 sec)

mysql> INSERT INTO `tl7154cbef` VALUES ('[0.902107]', 2011164482, '[0.828307,0.319142,0.696748,0.133167]', '[0.405775,0.866348,0.373082]');
Query OK, 1 row affected (0.01 sec)

mysql> WITH cte_331 (col_1805, col_1806, col_1807, col_1808, col_1809) AS (
    ->     SELECT 
    ->         /*+ read_from_storage(tiflash[tl7154cbef]) */ 
    ->         /*+ agg_to_cop() hash_agg() */ 
    ->         BIT_XOR(tl7154cbef.col_12) AS r0, 
    ->         GROUP_CONCAT(tl7154cbef.col_12 ORDER BY tl7154cbef.col_12) AS r1, 
    ->         CHARACTER_LENGTH(tl7154cbef.col_11) AS r2, 
    ->         COUNT(DISTINCT tl7154cbef.col_12) AS r3, 
    ->         REPLACE(tl7154cbef.col_13, tl7154cbef.col_11, tl7154cbef.col_13) AS r4 
    ->     FROM 
    ->         tl7154cbef 
    ->     WHERE 
    ->         tl7154cbef.col_13 IN ('[0.471552, 0.743432, 0.333821, 0.950423, 0.134729, 0.474538, 0.643419, 0.625898, 0.269346]') 
    ->     GROUP BY 
    ->         tl7154cbef.col_13, tl7154cbef.col_11 
    ->     ORDER BY 
    ->         r0, r1, r2, r3, r4 
    -> )
    -> SELECT 
    ->     1, col_1805, col_1806, col_1807, col_1808, col_1809 
    -> FROM 
    ->     cte_331 
    -> WHERE 
    ->     ISNULL(cte_331.col_1809) 
    -> ORDER BY 
    ->     1, 2, 3, 4, 5, 6 
    -> LIMIT 
    ->     21180486;
Empty set, 1 warning (0.10 sec)

mysql> explain WITH cte_331 (col_1805, col_1806, col_1807, col_1808, col_1809) AS (     SELECT          /*+ read_from_storage(tiflash[tl7154cbef]) */          /*+ agg_to_cop() hash_agg() */          BIT_XOR(tl7154cbef.col_12) AS r0,          GROUP_CONCAT(tl7154cbef.col_12 ORDER BY tl7154cbef.col_12) AS r1,          CHARACTER_LENGTH(tl7154cbef.col_11) AS r2,          COUNT(DISTINCT tl7154cbef.col_12) AS r3,          REPLACE(tl7154cbef.col_13, tl7154cbef.col_11, tl7154cbef.col_13) AS r4      FROM          tl7154cbef      WHERE          tl7154cbef.col_13 IN ('[0.471552, 0.743432, 0.333821, 0.950423, 0.134729, 0.474538, 0.643419, 0.625898, 0.269346]')      GROUP BY          tl7154cbef.col_13, tl7154cbef.col_11      ORDER BY          r0, r1, r2, r3, r4  ) SELECT      1, col_1805, col_1806, col_1807, col_1808, col_1809  FROM      cte_331  WHERE      ISNULL(cte_331.col_1809)  ORDER BY      1, 2, 3, 4, 5, 6  LIMIT      21180486;
+------------------------------------------+---------+--------------+------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                       | estRows | task         | access object    | operator info                                                                                                                                                                                                                                                                                                        |
+------------------------------------------+---------+--------------+------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Projection_15                            | 6.40    | root         |                  | 1->Column#26, Column#21, Column#22, Column#24, Column#23, Column#25                                                                                                                                                                                                                                                  |
| └─Projection_16                          | 6.40    | root         |                  | Column#21, Column#22, character_length(cast(test.tl7154cbef.col_11, var_string(1)))->Column#24, Column#23, replace(cast(test.tl7154cbef.col_13, var_string(5)), cast(test.tl7154cbef.col_11, var_string(1)), cast(test.tl7154cbef.col_13, var_string(5)))->Column#25                                                 |
|   └─Projection_27                        | 6.40    | root         |                  | Column#21, Column#22, Column#23, test.tl7154cbef.col_11, test.tl7154cbef.col_13                                                                                                                                                                                                                                      |
|     └─TopN_19                            | 6.40    | root         |                  | Column#21, Column#22, Column#35, Column#23, Column#36, offset:0, count:21180486                                                                                                                                                                                                                                      |
|       └─Projection_28                    | 6.40    | root         |                  | Column#21, Column#22, Column#23, test.tl7154cbef.col_11, test.tl7154cbef.col_13, character_length(cast(test.tl7154cbef.col_11, var_string(1)))->Column#35, replace(cast(test.tl7154cbef.col_13, var_string(5)), cast(test.tl7154cbef.col_11, var_string(1)), cast(test.tl7154cbef.col_13, var_string(5)))->Column#36 |
|         └─HashAgg_21                     | 6.40    | root         |                  | group by:Column#33, Column#34, funcs:bit_xor(Column#30)->Column#21, funcs:group_concat(Column#31 order by Column#30 separator ",")->Column#22, funcs:count(distinct Column#32)->Column#23, funcs:firstrow(Column#33)->test.tl7154cbef.col_11, funcs:firstrow(Column#34)->test.tl7154cbef.col_13                      |
|           └─Projection_26                | 8.00    | root         |                  | test.tl7154cbef.col_12->Column#30, cast(test.tl7154cbef.col_12, var_string(20))->Column#31, test.tl7154cbef.col_12->Column#32, test.tl7154cbef.col_11->Column#33, test.tl7154cbef.col_13->Column#34                                                                                                                  |
|             └─TableReader_25             | 8.00    | root         |                  | MppVersion: 2, data:ExchangeSender_24                                                                                                                                                                                                                                                                                |
|               └─ExchangeSender_24        | 8.00    | mpp[tiflash] |                  | ExchangeType: PassThrough                                                                                                                                                                                                                                                                                            |
|                 └─Selection_23           | 8.00    | mpp[tiflash] |                  | isnull(replace("[0.471552,0.743432,0.333821,0.950423,0.134729,0.474538,0.643419,0.625898,0.269346]", cast(test.tl7154cbef.col_11, var_string(1)), "[0.471552,0.743432,0.333821,0.950423,0.134729,0.474538,0.643419,0.625898,0.269346]"))                                                                             |
|                   └─TableFullScan_22     | 10.00   | mpp[tiflash] | table:tl7154cbef | pushed down filter:eq(test.tl7154cbef.col_13, [0.471552,0.743432,0.333821,0.950423,0.134729,0.474538,0.643419,0.625898,0.269346]), keep order:false, stats:pseudo                                                                                                                                                    |
+------------------------------------------+---------+--------------+------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
11 rows in set, 3 warnings (0.01 sec)
```


other case:
```sql
mysql> CREATE TABLE my_table (
    ->          `col_11` VARCHAR(255) NOT NULL,
    ->          `col_12` int(10) unsigned NOT NULL,
    ->          `col_13` VARCHAR(255) NOT NULL,
    ->          `col_14` VARCHAR(255) NOT NULL
    ->      );
Query OK, 0 rows affected (0.04 sec)

mysql> insert into my_table values ('World',1203945,'Hello, World!','FOR TEST');
Query OK, 1 row affected (0.00 sec)

mysql> ALTER TABLE my_table SET TIFLASH REPLICA 1;
Query OK, 0 rows affected (0.04 sec)

mysql> SELECT /*+ read_from_storage(tiflash[my_table]) */ REPLACE('Hello World', my_table.col_11, my_table.col_14)  FROM my_table;
+----------------------------------------------------------+
| REPLACE('Hello World', my_table.col_11, my_table.col_14) |
+----------------------------------------------------------+
| Hello FOR TEST                                            |
+----------------------------------------------------------+
1 row in set (0.09 sec)

mysql> explain SELECT /*+ read_from_storage(tiflash[my_table]) */ REPLACE('Hello World', my_table.col_11, my_table.col_14)  FROM my_table;
+---------------------------+----------+--------------+----------------+----------------------------------------------------------------------------+
| id                        | estRows  | task         | access object  | operator info                                                              |
+---------------------------+----------+--------------+----------------+----------------------------------------------------------------------------+
| TableReader_10            | 10000.00 | root         |                | MppVersion: 2, data:ExchangeSender_9                                       |
| └─ExchangeSender_9        | 10000.00 | mpp[tiflash] |                | ExchangeType: PassThrough                                                  |
|   └─Projection_4          | 10000.00 | mpp[tiflash] |                | replace(Hello World, test.my_table.col_11, test.my_table.col_14)->Column#6 |
|     └─TableFullScan_8     | 10000.00 | mpp[tiflash] | table:my_table | keep order:false, stats:pseudo                                             |
+---------------------------+----------+--------------+----------------+----------------------------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> SELECT /*+ read_from_storage(tiflash[my_table]) */ REPLACE('Hello World', my_table.col_11, 'forMYtest')  FROM my_table;
+------------------------------------------------------+
| REPLACE('Hello World', my_table.col_11, 'forMYtest') |
+------------------------------------------------------+
| Hello forMYtest                                       |
+------------------------------------------------------+
1 row in set (0.07 sec)

mysql> explain SELECT /*+ read_from_storage(tiflash[my_table]) */ REPLACE('Hello World', my_table.col_11, 'forMYtest')  FROM my_table;
+---------------------------+----------+--------------+----------------+-----------------------------------------------------------------+
| id                        | estRows  | task         | access object  | operator info                                                   |
+---------------------------+----------+--------------+----------------+-----------------------------------------------------------------+
| TableReader_10            | 10000.00 | root         |                | MppVersion: 2, data:ExchangeSender_9                            |
| └─ExchangeSender_9        | 10000.00 | mpp[tiflash] |                | ExchangeType: PassThrough                                       |
|   └─Projection_4          | 10000.00 | mpp[tiflash] |                | replace(Hello World, test.my_table.col_11, forMYtest)->Column#6 |
|     └─TableFullScan_8     | 10000.00 | mpp[tiflash] | table:my_table | keep order:false, stats:pseudo                                  |
+---------------------------+----------+--------------+----------------+-----------------------------------------------------------------+
4 rows in set (0.00 sec)




```

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- bugfix or new feature needs a release note -->

```release-note
None
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsStringReplace.h [hunk 1]
```diff
@@ -122,10 +122,40 @@ class FunctionStringReplace : public IFunction
 
         ColumnWithTypeAndName & column_result = block.getByPosition(result);
 
+        bool src_const = column_src->isColumnConst();
         bool needle_const = column_needle->isColumnConst();
         bool replacement_const = column_replacement->isColumnConst();
 
-        if (needle_const && replacement_const)
+        if (src_const){
```
Comment (breezewish) [inline dbms/src/Functions/FunctionsStringReplace.h:129]: Please fix the format.

File: dbms/src/Functions/FunctionsStringSearch.cpp [hunk 2]
```diff
@@ -1063,6 +1063,166 @@ struct ReplaceStringImpl
         }
     }
 
+    // Handle the case where `column_src` and `replace` are const
+    static void vectorConstReplacement(
```
Comment (breezewish) [inline dbms/src/Functions/FunctionsStringSearch.cpp:1067]: Can we modify existing functions? Existing functions are already capable of handling a list of sources. It should be better to allow it handling constant source (i.e. one source), minimizing the changes.

File: dbms/src/Functions/FunctionsStringSearch.cpp [hunk 3]
```diff
@@ -1063,6 +1063,166 @@ struct ReplaceStringImpl
...
+
+                if (match == data.data() + data.size())
+                {
+                    break;
+                }
+
+                if (pos_in_data + needle_size <= data.size())
+                {
+                    auto replacement_offset = StringUtil::offsetAt(replacement_offsets, i);
+                    auto replacement_size = StringUtil::sizeAt(replacement_offsets, i) - 1;  // 忽略末尾的零字节
```
Comment (breezewish) [inline dbms/src/Functions/FunctionsStringSearch.cpp:1203]: Don't use chinese.

File: dbms/src/Functions/FunctionsStringReplace.h [hunk 4]
```diff
@@ -122,10 +122,40 @@ class FunctionStringReplace : public IFunction
...
+                    occ,
+                    match_type,
+                    column_result
+                );
+            }else
+            {
+                throw Exception(
+                    "UnImplement function.",
+                    ErrorCodes::BAD_ARGUMENTS);
+            }
```
Comment (breezewish) [inline dbms/src/Functions/FunctionsStringReplace.h:157]: Possibly no need for a lot of impls because we are fixing for an edge case that only discovered in tests. It is acceptable to be not performance optimal.

File: dbms/src/Functions/FunctionsStringSearch.cpp (hunk 2 above)
Comment (EricZequan) [reply dbms/src/Functions/FunctionsStringSearch.cpp:1067]: I think these functions are necessary because the first parameter of existing functions is `Column` by default, which will be parsed into `ColumnString` in `FunctionsStringSearch.cpp`, and our implementation needs to parse the parameter into `ColumnConst`. `ColumnConst` does not have `getChars` and `getOffsets` methods, so existing functions cannot be used.

File: dbms/src/Functions/FunctionsStringReplace.h (hunk 4 above)
Comment (EricZequan) [reply dbms/src/Functions/FunctionsStringReplace.h]: In my test case, some sql like :
```sql
explain SELECT /*+ read_from_storage(tiflash[my_table]) */ REPLACE('Hello World', my_table.col_11, my_table.col_14)  FROM my_table;
```
will get same error.  I think it may need to process, so I add these function to deal with such problems. 😂

File: dbms/src/Functions/FunctionsStringSearch.cpp (hunk 2 above)
Comment (breezewish) [reply dbms/src/Functions/FunctionsStringSearch.cpp:1067]: It doesn't make sense. Think about it. You can now calculate a bunch of FN(A, B), why cannot calculate a single row of FN(ConstA, ConstB)?

Comment (EricZequan) [reply dbms/src/Functions/FunctionsStringSearch.cpp]: Yes, I found a way to convert between the two types, so that now the existing function can be used directly, and the amount of code is greatly reduced.

File: dbms/src/Functions/FunctionsStringReplace.h [hunk 5]
```diff
@@ -280,7 +232,28 @@ class FunctionStringReplace : public IFunction
             const auto * col_replacement_const = typeid_cast<const ColumnConst *>(column_replacement.get());
             auto replacement = col_replacement_const->getValue<String>();
 
-            if (const auto * col = checkAndGetColumn<ColumnString>(column_src.get()))
+            bool col_const = column_src->isColumnConst();
+
+            if (col_const)
+            {
+                auto new_src = column_src->convertToFullColumnIfConst();
```
Comment (breezewish) [inline dbms/src/Functions/FunctionsStringReplace.h:251]: This cost is too high, why not simply provide a reference?

File: dbms/src/Functions/FunctionsStringReplace.h [hunk 6]
```diff
@@ -280,7 +232,28 @@ class FunctionStringReplace : public IFunction
             const auto * col_replacement_const = typeid_cast<const ColumnConst *>(column_replacement.get());
             auto replacement = col_replacement_const->getValue<String>();
 
-            if (const auto * col = checkAndGetColumn<ColumnString>(column_src.get()))
+            bool col_const = column_src->isColumnConst();
```
Comment (breezewish) [inline dbms/src/Functions/FunctionsStringReplace.h:247]: Use Const Generics to avoid these logics (and costs) are involved in other branches (i.e. non-const).

**Review by gengliqi — COMMENTED** [review]
It's better to add a const data version of `vectorNonConstNeedle`, `vectorNonConstReplacement`, and `vectorNonConstNeedleReplacement`.

File: dbms/src/Functions/FunctionsStringReplace.h (hunk 5 above)
Comment (EricZequan) [reply dbms/src/Functions/FunctionsStringReplace.h]: here i use:
```cpp
// using the data directly as a reference.
const auto & const_data = col_const->getDataColumn();
const auto * col = typeid_cast<const ColumnString *>(&const_data);
```
instad to avoid these cost now.

File: dbms/src/Functions/FunctionsStringReplace.h (hunk 6 above)
Comment (EricZequan) [reply dbms/src/Functions/FunctionsStringReplace.h]: i changed it to: 
```cpp
if (const auto * col_const = checkAndGetColumnConst<ColumnString>(column_src.get()))
```

**Comment by EricZequan** [comment]
> It's better to add a const data version of `vectorNonConstNeedle`, `vectorNonConstReplacement`, and `vectorNonConstNeedleReplacement`.
To keep the code non-redundant, maybe we can not add it? 🤔

The processing logic of the `constant` type function is the same as the original function, just need to interpret `constant` as `Column` to process it, this change only needs a few dozen lines.

by the way, I have tried this method in the `debug` commit and `add 2 function to complete replace for constant`
commit, and the overall code is not elegant enough

File: dbms/src/Functions/FunctionsStringReplace.h [hunk 7]
```diff
@@ -232,7 +242,29 @@ class FunctionStringReplace : public IFunction
             const auto * col_replacement_const = typeid_cast<const ColumnConst *>(column_replacement.get());
             auto replacement = col_replacement_const->getValue<String>();
 
-            if (const auto * col = checkAndGetColumn<ColumnString>(column_src.get()))
+            if (const auto * col_const = checkAndGetColumnConst<ColumnString>(column_src.get()))
+            {
+                // using the data directly as a reference.
+                const auto & const_data = col_const->getDataColumn();
+                const auto * col = typeid_cast<const ColumnString *>(&const_data);
```
[the remark this answers is not in the document]
Comment (EricZequan) [reply dbms/src/Functions/FunctionsStringReplace.h:249]: right. Now I add `vectorConstSrcAndReplace`、`vectorConstSrcAndNeedle`、`vectorConstSrc` to handle the corresponding case to ensure the same number of rows.

File: dbms/src/Functions/FunctionsStringSearch.cpp [hunk 8]
```diff
@@ -972,6 +972,225 @@ struct ReplaceStringImpl
         }
     }
 
+    static void vectorConstSrcAndReplace(
+        const ColumnString::Chars_t & data,
+        const ColumnString::Offsets & offsets,
```
[the remark this answers is not in the document]
Comment (EricZequan) [reply dbms/src/Functions/FunctionsStringSearch.cpp]: Updated. PTAL~

File: dbms/src/Functions/tests/gtest_strings_replace.cpp [hunk 9]
```diff
@@ -104,6 +104,38 @@ try
             toVec({"  hello   ", "   h e llo", "hello    ", "     ", "hello, world"}),
             toVec({" ", "h", "", "h", ","}),
             toVec({"", "x", "xx", " ", ","})));
+
+    /// const src replacement
+    ASSERT_COLUMN_EQ(
+        toVec({"Good Night", "Bad Afternoon", "Good Afterwhile"}),
+        executeFunction(
+            "replaceAll",
+            toVec({"Good Afternoon"}),
```
Comment (breezewish) [inline dbms/src/Functions/tests/gtest_strings_replace.cpp:113]: it should be `toConst`?

File: dbms/src/Functions/tests/gtest_strings_replace.cpp [hunk 10]
```diff
@@ -104,6 +104,38 @@ try
...
+            toVec({"Afternoon", "Good", "noon"}),
+            toVec({"Night", "Bad", "while"})));
+
+    /// const src and needle replacement
+    ASSERT_COLUMN_EQ(
+        toVec({"Good Night", "Good Bad", "Good while"}),
+        executeFunction(
+            "replaceAll",
+            toVec({"Good Afternoon"}),
```
Comment (breezewish) [inline dbms/src/Functions/tests/gtest_strings_replace.cpp:122]: it should be `toConst`?

File: dbms/src/Functions/tests/gtest_strings_replace.cpp [hunk 11]
```diff
@@ -104,6 +104,38 @@ try
...
+            toConst({"Afternoon"}),
+            toVec({"Night", "Bad", "while"})));
+
+    /// const src and replace replacement
+    ASSERT_COLUMN_EQ(
+        toVec({"Good Night", "Night Afternoon", "Good AfterNight"}),
+        executeFunction(
+            "replaceAll",
+            toVec({"Good Afternoon"}),
```
Comment (breezewish) [inline dbms/src/Functions/tests/gtest_strings_replace.cpp:131]: it should be `toConst`?

File: dbms/src/Functions/FunctionsStringSearch.cpp [hunk 12]
```diff
@@ -985,20 +984,25 @@ struct ReplaceStringImpl
         ColumnString::Chars_t & res_data,
         ColumnString::Offsets & res_offsets)
     {
-        res_data.reserve(data.size() * needle_offsets.size()); // Leave enough space to handle multiple lines
-        res_offsets.resize(needle_offsets.size());
+        auto data_col = ColumnString::create();
+        data_col->insert(data);
```
Comment (breezewish) [inline dbms/src/Functions/FunctionsStringSearch.cpp:988]: Don't do this, this greatly reduce the performance.

**Comment by gengliqi** [comment]
> > I'm still confused by why do we need almost 300 lines in order to make something work for a subset? It is wired.
> 
> I understand what you mean. Constant can actually be seen as a cloumn with only one line. It is reasonable to reuse the original function as a subset. However, liqi suggested adding a separate function to process. Emmm anyway, I will discuss it with him tomorrow~

What I mean isn't to say that separate functions should be used, you can use templates to reduce code repetition.

**Comment by breezewish** [comment]
I created a new PR here: https://github.com/pingcap/tiflash/pull/9536

**Comment by JaySon-Huang** [comment]
Fixed by https://github.com/pingcap/tiflash/pull/9615
